docs/coreneuron/compatibility.rst

```diff
@@ -0,0 +1,149 @@
+
+Some useful instructions to make a mod file thread safe are:
+
+* Make sure that all the variables that are written in the mod file and are defined as ``GLOBAL`` are now defined as ``RANGE``. More information about ``GLOBAL`` and ``RANGE`` variables can be found `here <https://nrn.readthedocs.io/en/latest/hoc/modelspec/programmatic/mechanisms/nmodl2.html>`_.
+* In case there are any ``VERBATIM`` blocks those need to be threadsafe and there must be a ``THREADSAFE`` keyword in the ``NEURON`` block to declare that the ``VERBATIM`` blocks are thread safe. Also make sure that the ``NEURON`` block is defined before any ``VERBATIM`` block in the mod file
+
+
+BBCOREPOINTER
+*************
+``BBCOREPOINTER`` is used to transfer dynamically allocated data between NEURON and CoreNEURON.
+
+User-allocated data can be managed in NMODL using the ``POINTER`` type.
+It allows the programmer to reference data that has been allocated in HOC or in ``VERBATIM`` blocks.
+This allows for more advanced data-structures that are not natively supported in NMODL.
+
+Since NEURON itself has no knowledge of the layout and size of this data it cannot
+transfer ``POINTER`` data automatically to CoreNEURON.
+Furtheremore, in many cases there is no need to transfer the data between the two instances.
+In some cases, however, the programmer would like to transfer certain user-defined data into CoreNEURON.
+The most prominent example are Random123 random number stream parameters used in synapse mechanisms.
+To support this use-case the ``BBCOREPOINTER`` type was introduced.
+Variables that are declared as ``BBCOREPOINTER`` behave exactly the same as ``POINTER`` but are
+additionally taken into account when NEURON is serializing mechanism data (for file writing or
+direct-memory transfer).
+For NEURON to be able to write (and indeed CoreNEURON to be able to read) ``BBCOREPOINTER``
+data, the programmer has to additionally provide two C functions that are called as part
+of the serialization/deserialization.
+
+.. code-block:: c++
+
+   static void bbcore_write(double* x, int* d, int* d_offset, int* x_offset, _threadargsproto_);
+   static void bbcore_read(double* x, int* d, int* d_offset, int* x_offset, _threadargsproto_);
+
+The implementation of ``bbcore_write`` and ``bbcore_read`` determines the serialization and
+deserialization of the per-instance mechanism data referenced through the various
+``BBCOREPOINTER``.
+
+NEURON will call ``bbcore_write`` twice per mechanism instance.
+In a first sweep, the call is used to determine the required memory to be allocated on the serialization arrays.
+In the second sweep the call is used to fill in the data per mechanism instance.
+
+The functions take following arguments
+
+* ``x``: A ``double`` type array that will be allocated by NEURON to fill with real-valued data. In the
```

Comment:
Done; in the slightly longer term it would be nice not to do this manually, although these callbacks are a bit unusual to document (I guess https://github.com/BlueBrain/CoreNeuron/blob/43e881eb55f9191dba3c3db797b498bf60ccce96/coreneuron/mechanism/membfunc.hpp#L144-L154 is the relevant bit of code 🤷)